compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp

```diff
@@ -0,0 +1,217 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <functional>
+#include "iree/compiler/Codegen/Common/GPU/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/GPU/TransformOps/GPUTransformOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/IR/DeviceMappingInterface.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPUDISTRIBUTEFORALLPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+
+struct GPUDistributeForallPass final
+    : impl::GPUDistributeForallPassBase<GPUDistributeForallPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+LogicalResult resolveGPUMappedForallOp(RewriterBase &rewriter,
+                                       scf::ForallOp forallOp,
+                                       Value linearThreadId,
+                                       int64_t flatWorkgroupSize,
+                                       int64_t subgroupSize) {
+
+  // Skip forall ops without mappings.
+  if (!forallOp.getMapping()) {
+    return success();
+  }
+
+  ArrayAttr mapping = forallOp.getMappingAttr();
+  bool hasThreadMapping =
+      llvm::all_of(mapping, llvm::IsaPred<gpu::GPUThreadMappingAttr>);
+  bool hasWarpMapping =
+      llvm::all_of(mapping, llvm::IsaPred<gpu::GPUWarpMappingAttr>);
+
+  // Skip forall ops that are not mapped to GPU ids.
+  if (!hasThreadMapping && !hasWarpMapping) {
+    return success();
+  }
+
+  if (forallOp->getNumResults() != 0) {
+    forallOp.emitOpError("Cannot distribute scf.forall op on tensors.");
+    return failure();
+  }
+
+  if (!isDescendingRelativeMappingIndices(mapping.getValue())) {
+    forallOp.emitOpError("Cannot distribute forall op with non-descending "
+                         "relative iterator mapping");
+    return failure();
+  }
+
+  if (!llvm::all_of(mapping, [](Attribute attr) {
+        return cast<DeviceMappingAttrInterface>(attr).isLinearMapping();
+      })) {
+    forallOp.emitOpError("unimplemented: resolution of scf.forall ops without "
+                         "linear id mappings.");
+    return failure();
+  }
+
+  if (!forallOp.isNormalized()) {
```

Comment:
I was thinking we could just add the loop normalization pass immediately before this rather than complicating the code here. Either way is ok with me though.